designs/SpanMarshallers.md

```diff
@@ -0,0 +1,262 @@
+# Support for Marshalling `(ReadOnly)Span<T>`
+
+As part of the exit criteria for the DllImportGenerator experiment, we have decided to introduce support for marshalling `System.Span<T>` and `System.ReadOnlySpan<T>` into the DllImportGenerator-generated stubs. This document describes design decisions made during the implementation of these marshallers.
+
+## Design 1: "Intrinsic" support for `(ReadOnly)Span<T>`
+
+In this design, the default support for `(ReadOnly)Span<T>` is emitted into the marshalling stub directly and builds on the pattern we enabled for arrays.
+
+### Default behavior
+
+This section describes the behavior of the default emitted-in-stub marshallers.
+
+By default, we will marshal `Span<T>` and `ReadOnlySpan<T>` similarly to array types. When possible, we will pin the `(ReadOnly)Span<T>`'s data and pass that down. When it is not possible to do so, we will try stack allocating scratch space for efficiency or allocate native memory with `Marshal.AllocCoTaskMem` for compatibility with arrays.
+
+To support marshalling from native to managed, we will support the same `MarshalAsAttribute` properties that arrays support today.
+
+When a `(ReadOnly)Span<T>` is marshalled from native to managed, we will allocate a managed array and copy the data from the native memory into the array.
```

Comment:
>we will allocate a managed array and copy the data from the native memory into the array.

This is only in the case where the managed callee has a `T[]` as opposed to a `Span<T>` right? I would assume the native caller would provide stable memory and if passes in a `int*,int` we can convert that to a `Span<int>` without allocating anything.

Reply:
The default marshalling has to handle the blittable and non-blittable cases. The non-blittable case cannot use the native memory directly, so we can't use that in the default case. This proposal has the non-allocating `Span<T>` as an opt-in marshaller.